src/executables/csql.c

```diff
@@ -1460,6 +1464,205 @@ csql_do_session_cmd (char *line_read, CSQL_ARGUMENT * csql_arg)
+	    {
+	      *(host_name - 1) = '\0';
+	    }
+	}
+
+      strcpy (dbname_cpy, argument);
+      strcpy (username_cpy, argument);
+      /*find db name following the user name */
+      db_name = strrchr (dbname_cpy, ' ');
+      if (db_name != NULL)
+	{
+	  if (*(db_name + 1) == '\0')
+	    {
+	      db_name = NULL;
+	      csql_Error_code = CSQL_ERR_CONNECT;
+	      return DO_CMD_FAILURE;
+	    }
+	  else
+	    {
+	      db_name += 1;
+	    }
+	}
+      user_name = strtok_r (username_cpy, delim, &save_ptr_strtok);
+      if (user_name == NULL)
+	{
+	  csql_Error_code = CSQL_ERR_CONNECT;
+	  return DO_CMD_FAILURE;
+
+	}
+
+      memset (&csql_new_arg, 0, sizeof (CSQL_ARGUMENT));
+      csql_arg_copy (&csql_new_arg, csql_arg);
+
+      /*connect to same DB user or not */
+      if ((db_name == NULL || strcasecmp (csql_arg->db_name, db_name) == 0) && csql_Database_connected == TRUE)
+	{
+
+	  if (!au_is_dba_group_member (Au_user))
+	    {
+
+	      error_code = au_login (user_name, NULL, false);
+	      if (csql_Is_interactive && db_error_code () == ER_AU_INVALID_PASSWORD)
+		{
+		  p = getpass ((char *) csql_get_message (CSQL_PASSWD_PROMPT_TEXT));
+		  error_code = au_login (user_name, p, false);
+		}
+	    }
+	  else
+	    {
+	      error_code = au_login (user_name, NULL, false);
+	    }
+	  if (error_code == NO_ERROR)
+	    {
+	      error_code = clogin_user (user_name);
+	    }
+	  else
+	    {
+	      return DO_CMD_FAILURE;
+	    }
+	  if (error_code == NO_ERROR)
+	    {
+	      csql_new_arg.user_name = strdup (user_name);
+	      csql_new_arg.db_name = strdup (csql_arg->db_name);
+	      if (p == NULL || p[0] == '\0')
+		{
+		  csql_new_arg.passwd = (char *) NULL;	/* to fit into db_login protocol */
+		}
+	      else
+		{
+		  csql_new_arg.passwd = strdup (p);
+		}
+
+	      memcpy (csql_arg, &csql_new_arg, sizeof (CSQL_ARGUMENT));
+	      fprintf (csql_Output_fp, "CONNECTED : %s %s\n", csql_arg->user_name, csql_arg->db_name);
+	    }
+
+	}
+      else
+	{
+#if defined(CS_MODE)
+	  memset (boot_Host_connected, 0, sizeof (boot_Host_connected));
+#endif /* CS_MODE */
+
+	  /*Failed to access other host or db and then access formal db_name */
+	  if (db_name == NULL)
+	    {
+	      db_name = strdup (csql_arg->db_name);
+	    }
+
+	  if (csql_Database_connected)
+	    {
+	      csql_Database_connected = false;
+	      db_end_session ();
+	      db_shutdown ();
+	    }
+	  er_init ("./csql.err", ER_NEVER_EXIT);
+	  if (db_restart_ex (UTIL_CSQL_NAME, db_name, user_name, NULL, NULL, db_get_client_type ()) != NO_ERROR)
+	    {
+	      p = getpass ((char *) csql_get_message (CSQL_PASSWD_PROMPT_TEXT));
+
+	      /* try again */
+	      if (db_restart_ex (UTIL_CSQL_NAME, db_name, user_name, p, NULL, db_get_client_type ()) != NO_ERROR)
+		{
+		  csql_Error_code = CSQL_ERR_SQL_ERROR;
+		  csql_check_server_down ();
+
+		  fprintf (csql_Output_fp, "SQL session is disconnected\n");
```

Comment:
fprintf (csql_Output_fp, "Warning: current CSQL session is disconnected.\n");

Reply:
Ok. thank you.